### PR TITLE
feat/restore-snapshot

### LINKED ISSUE #6: model / custom node provisioning script
If we could somehow define a 'provisioning' type script that allows for custom models etc to be added to the container image that'd be great! I seem to be having severe issues when trying to manually re-create the image locally :( 

Perhaps something similar to the ai-dock runpod images (which is set via an env variable pointing to a github url where the prov script is, see the docker-compose for the project): 
https://github.com/ai-dock/comfyui/tree/main/config/provisioning

edit: not really sure if this is possible for during the image build step, might still need to rebuild the image for custom models, but it'd be nice! :D

### PR DESCRIPTION
**Pull request by kklemon**
## Motivation

Many users and workflows rely on custom nodes and extensions which correspondingly should also be available in a RunPod worker.

[ComfyUI Manager](https://github.com/ltdrdata/ComfyUI-Manager), perhaps the most popular ComfyUI management exension, provides a [convenient feature](https://github.com/ltdrdata/ComfyUI-Manager?tab=readme-ov-file#snapshot-manager) to export a snapshot of the ComfyUI instance with all installed extensions.

This PR implements optional restoring of a snapshot if provided as `snapshot.json` in the root directory. If such a file is found, ComfyUI Manager will be installed and restoring of the snapshot will be triggered by starting ComfyUI with the `--cpu --quick-test-for-ci` flags within the Docker image build process.

## Issues closed

#6 #22 #18 #9 (partially)

## Todo

- [ ] Implement tests after this initial proposal is approved.

### CODE REVIEW

**Comment by kklemon** [comment]
Adding a `snapshot.json` in the root directory should be sufficient. Also, make sure that you are building the image from scratch. Apart from these suggestions, I would need to take a look at the build logs to provide more help.

**Comment by triplecookedchips** [comment]
Thanks @kklemon - I didn't build the image from scratch, so will try again with your advice

File: Dockerfile [hunk 1]
```diff
@@ -39,15 +41,16 @@ RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/vae/sdxl_vae.safetenso
 RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/vae/sdxl-vae-fp16-fix.safetensors https://huggingface.co/madebyollin/sdxl-vae-fp16-fix/resolve/main/sdxl_vae.safetensors; fi
 RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/loras/xl_more_art-full_v1.safetensors https://civitai.com/api/download/models/152309; fi
 
-# Support for the network volume
-ADD src/extra_model_paths.yaml ./
```
Comment (TimPietrusky) [inline Dockerfile:43]: Why is this removed? As this is also usable for adding models into ComfyUI via the network volume.

File: Dockerfile [hunk 2]
```diff
@@ -39,15 +41,16 @@ RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/vae/sdxl_vae.safetenso
...
 
-# Add the start and the handler
-ADD src/start.sh src/rp_handler.py test_input.json ./
-RUN chmod +x /start.sh
+# Add scripts
+ADD src/start.sh src/restore_snapshot.sh src/rp_handler.py test_input.json ./
+RUN chmod +x /start.sh /restore_snapshot.sh
+
+# Optionally copy snapshot file
+ADD snapshot.jso[n] /
```
Comment (TimPietrusky) [inline Dockerfile:51]: Why would we have `jso[n]` here when we always expect a json-file?

**Review by TimPietrusky — CHANGES_REQUESTED** [review]
I have two questions, please take a look!

**Comment by TimPietrusky** [comment]
@kklemon I updated the base branch as we want to move away from having the `dev` branch.

**Comment by martintomov** [comment]
I have around 30 custom nodes - what's the best way to install nodes right now? Using the Network Volume where we store models?

**Comment by kklemon** [comment]
@TimPietrusky I've completely forgotten about this PR. I will check whether it's still relevant, update it if that's the case or close it otherwise.

**Comment by kklemon** [comment]
@martintomov I haven't been active in the ComfyUI community for a while and don't know what the current _de facto_ standard for managing extensions is, but when I created this PR, I assumed that it's [ComfyUI Manager](https://github.com/ltdrdata/ComfyUI-Manager).

If you want to incorporate custom nodes into a RunPod deployment, my recommendation is to install your nodes or extensions via ComfyUI Manager and [export a snapshot of your setup as JSON file](https://github.com/ltdrdata/ComfyUI-Manager?tab=readme-ov-file#snapshot-manager). You can then just put the `snapshot.json` in the root directory of this repository, which will be recognized by the build script when building the Docker image. The extensions will be installed during the build process.

**Comment by martintomov** [comment]
@kklemon, thanks for your recommendation. I forked your PR, made a `snapshot.json` export from ComfyUI Manager with a `test_input.json` workflow, but I faced unhealthy workers and no execution. Any tips that you can give on this? I tried with a pretty simple workflow too, just to get started with it.

Before that, I found something that you could change in the dockerfile, as an error is occurring because the script can't find the python command - it's trying to use python but in the Dockerfile we install python3.10:
```
# Install Python, git and other necessary tools
RUN apt-get update && apt-get install -y \
 python3.10 \
 python3-pip \
 git \
 wget \
 && ln -sf /usr/bin/python3.10 /usr/bin/python  # Added this line
 ```
 My dockerfile looks like this at the end:
 ```
 # Use Nvidia CUDA base image
FROM nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu22.04 as base

# Prevents prompts from packages asking for user input during installation
ENV DEBIAN_FRONTEND=noninteractive
# Prefer binary wheels over source distributions for faster pip installations
ENV PIP_PREFER_BINARY=1
# Ensures output from python is printed immediately to the terminal without buffering
ENV PYTHONUNBUFFERED=1 
# Speed up some cmake builds
ENV CMAKE_BUILD_PARALLEL_LEVEL=8

# Install Python, git and other necessary tools
RUN apt-get update && apt-get install -y \
    python3.10 \
    python3-pip \
    git \
    wget \
    && ln -sf /usr/bin/python3.10 /usr/bin/python

# Clean up to reduce image size
RUN apt-get autoremove -y && apt-get clean -y && rm -rf /var/lib/apt/lists/*

# Clone ComfyUI repository
RUN git clone https://github.com/comfyanonymous/ComfyUI.git /comfyui

# Change working directory to ComfyUI
WORKDIR /comfyui

# Install ComfyUI dependencies
RUN pip3 install --no-cache-dir torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121 \
    && pip3 install --no-cache-dir xformers==0.0.21 \
    && pip3 install -r requirements.txt

# Install runpod
RUN pip3 install runpod requests

ARG SKIP_DEFAULT_MODELS
# Download checkpoints/vae/LoRA to include in image.
RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/unet/flux1-schnell-fp8-e4m3fn.safetensors https://huggingface.co/Kijai/flux-fp8/resolve/main/flux1-schnell-fp8-e4m3fn.safetensors; fi
RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/vae/flux-vae-bf16.safetensors https://huggingface.co/Kijai/flux-fp8/resolve/main/flux-vae-bf16.safetensors; fi
RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/clip/clip_l.safetensors https://huggingface.co/comfyanonymous/flux_text_encoders/resolve/main/clip_l.safetensors; fi
RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/clip/t5xxl_fp16.safetensors https://huggingface.co/comfyanonymous/flux_text_encoders/resolve/main/t5xxl_fp16.safetensors; fi
RUN if [ -z "$SKIP_DEFAULT_MODELS" ]; then wget -O models/loras/mrr-synthetic-data-experiment-v3-schnell.safetensors https://huggingface.co/martintomov/mrr-synthetic-data-experiment-v3-schnell/resolve/main/mrr-synthetic-data-experiment-v3-schnell.safetensors; fi

WORKDIR /

# Add scripts
ADD src/start.sh src/restore_snapshot.sh src/rp_handler.py test_input.json ./
RUN chmod +x /start.sh /restore_snapshot.sh

# Optionally copy snapshot file
ADD snapshot.jso[n] /

RUN /restore_snapshot.sh

# Start the container
CMD /start.sh
```
while I also updated `extra_model_paths.yaml`, but no positive effect:
```
runpod_worker_comfy:
  base_path: /runpod-volume
  checkpoints: models/checkpoints/
  clip: models/clip/
  clip_vision: models/clip_vision/
  configs: models/configs/
  controlnet: models/controlnet/
  embeddings: models/embeddings/
  loras: models/loras/
  upscale_models: models/upscale_models/
  vae: models/vae/
  unet: models/unet/
```

What am i missing here? The repo's template `flux-schnell` and `flux-dev` containers are working fine.

File: src/restore_snapshot.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+pip install -r requirements.txt
+
+mkdir startup-scripts
+mv /snapshot.json startup-scripts/restore-snapshot.json
+
+cd ../..
+
+# Trigger restoring of the snapshot by performing a quick test run
+# Note: We need to use `yes` as some custom nodes may try to install dependencies with pip
+/usr/bin/yes | python main.py --cpu --quick-test-for-ci
```
Comment (Sumukh) [inline src/restore_snapshot.sh:28]: Does this need to be python3?

Comment (martintomov) [reply src/restore_snapshot.sh:28]: I think it's a good idea to add a symlink for python to point to python3.10. 

Here is the relevant section of the Dockerfile with the symlink added:
```
RUN apt-get update && apt-get install -y \
 python3.10 \
 python3-pip \
 git \
 wget \
 && ln -sf /usr/bin/python3.10 /usr/bin/python
```

**Comment by TimPietrusky** [comment]
@kklemon thank you very much, then I will go down this road and make sure it is working.

**Comment by TimPietrusky** [comment]
Some changes are coming: 

* added comfy-cli
* install ComfyUI with comfy-cli
* restore a snapshot using comfy-cli
* update docs

**Comment by A-BMT02** [comment]
I can't seem to get this working . When i start a runpod pod and open my comfy ui on port 8188 it works fine , i can generate my images fine 

But when i use the serverless endpoint with the same workflow json , i get an error 
`2024-11-19 14:00:45.074
[c7msa02ut39p9w]
[info]
invalid prompt: {'type': 'invalid_prompt', 'message': 'Cannot execute because node FaceDetailer does not exist.', `

It seems like the custom nodes are not being recognized in the serverless endpoint even though i have the snapshot.json in the /runpod-volume (/workspace) directory


If i try a request with no custom nodes on the runpod serverless endpoint , it works fine.

**Comment by TimPietrusky** [comment]
@A-BMT02 I moved this conversation over into an issue: https://github.com/blib-la/runpod-worker-comfy/issues/79
